client/state/state_database.go

```diff
@@ -13,51 +16,75 @@ import (
 /*
 The client has a boltDB backed state store. The schema as of 0.9 looks as follows:
 
-allocations/ (bucket)
-|--> <alloc-id>/ (bucket)
-    |--> alloc -> *structs.Allocation
-    |--> alloc_runner persisted objects (k/v)
-	|--> <task-name>/ (bucket)
-        |--> task_runner persisted objects (k/v)
-
-devicemanager/
-|--> plugin-state -> *dmstate.PluginState
+meta/
+|--> version -> "1"
+|--> upgraded -> time.Now().Format(timeRFC3339)
+allocations/
+|--> <alloc-id>/
+   |--> alloc         -> allocEntry{*structs.Allocation}
+   |--> deploy_status -> deployStatusEntry{*structs.AllocDeploymentStatus}
+   |--> task-<name>/
+      |--> local_state -> *trstate.LocalState # Local-only state
+      |--> task_state  -> *structs.TaskState  # Sync'd to servers
+
+devices/
+|--> plugin_state -> *dmstate.PluginState
 */
 
 var (
+	// metaBucketName is the name of the metadata bucket
+	metaBucketName = []byte("meta")
+
+	// metaVersionKey is the key the state schema version is stored under.
+	metaVersionKey = []byte("version")
+
+	// metaVersion is the value of the state schema version to detect when
+	// an upgrade is needed. It skips the usual boltdd/msgpack backend to
+	// be as portable and futureproof as possible.
+	metaVersion = "1"
+
+	// metaUpgradedKey is the key that stores the timestamp of the last
+	// time the schema was upgraded.
+	metaUpgradedKey = []byte("upgraded")
+
 	// allocationsBucketName is the bucket name containing all allocation related
 	// data
 	allocationsBucketName = []byte("allocations")
 
-	// allocKey is the key serialized Allocations are stored under
+	// allocKey is the key Allocations are stored under encapsulated in
+	// allocEntry structs.
 	allocKey = []byte("alloc")
 
-	// taskRunnerStateAllKey holds all the task runners state. At the moment
-	// there is no need to split it
-	//XXX Old key - going to need to migrate
-	//taskRunnerStateAllKey = []byte("simple-all")
+	// allocDeployStatusKey is the key *structs.AllocDeploymentStatus is
+	// stored under.
+	allocDeployStatusKey = []byte("deploy_status")
 
-	// allocations -> $allocid -> $taskname -> the keys below
+	// allocations -> $allocid -> task-$taskname -> the keys below
 	taskLocalStateKey = []byte("local_state")
 	taskStateKey      = []byte("task_state")
 
 	// devManagerBucket is the bucket name containing all device manager related
 	// data
-	devManagerBucket = []byte("devicemanager")
+	devManagerBucket = []byte("devices")
```

Comment:
Be aware of conflicts here if this is merged first: https://github.com/hashicorp/nomad/pull/4964/files#diff-64749d8d8c42d5b993eff64d7f522d00

Reply:
I'll revert my change.